Mathlib/RingTheory/Kaehler.lean

```diff
@@ -511,6 +511,7 @@ theorem KaehlerDifferential.kerTotal_mkQ_single_algebraMap_one (x) : (x𝖣1) =
 #align kaehler_differential.ker_total_mkq_single_algebra_map_one KaehlerDifferential.kerTotal_mkQ_single_algebraMap_one
 
 theorem KaehlerDifferential.kerTotal_mkQ_single_smul (r : R) (x y) : (y𝖣r • x) = r • y𝖣x := by
+  letI : SMulZeroClass R S := inferInstance
   rw [Algebra.smul_def, KaehlerDifferential.kerTotal_mkQ_single_mul,
     KaehlerDifferential.kerTotal_mkQ_single_algebraMap, add_zero, ← LinearMap.map_smul_of_tower,
     Finsupp.smul_single, mul_comm, Algebra.smul_def]
```

Comment:
This regression is strange. Somehow after `rw [← LinearMap.map_smul_of_tower]`, Lean suddenly needs to be reminded of the SMulZeroClass instance, otherwise it can't `rw Finsupp.smul_single`. `map_smul_of_tower` consumes a `LinearMap.CompatibleSMul` instance, so maybe due to the new instances introduced, the rw produces a syntactically slightly different result. I think it's a pretty well-known bug that Lean has trouble inferring weaker instances from stronger ones, e.g. Algebra to Module to SMul. (Maybe it matters whether the stronger `extend`s the weaker one?)

Reply:
Does it timeout or fail?

Reply:
I think it's a fail (says it can't synthesize `SMulZeroClass R S`)

Reply:
That is strange. The well-known bug (that I know of) just times out.

Reply:
I think the three instance reminders [here](https://github.com/leanprover-community/mathlib4/blob/8aa9689f0e48c6fcf9b3ad6846b722c13b7fcd53/Mathlib/RingTheory/IntegralClosure.lean#L855-L862) are all failures.

Reply:
The comment says it times out though. Anyways I'm still waiting for the cache so I can check it out myself.

Reply:
Sorry, I just checked and you're right; all those three instances come with a
> (deterministic) timeout at 'typeclass', maximum number of heartbeats (20000) has been reached (use 'set_option synthInstance.maxHeartbeats <num>' to set the limit)

message, while this timeout in Kaehler does not.

Reply:
I just compared the context and the goal state after `rw [← LinearMap.map_smul_of_tower]` on master against this PR, with pp.all turned on. It turns out they are completely identical, so I'm puzzled why Lean suddenly needs the reminder ...

Reply:
I looked into it myself and it seems like some strange bug in the synthesizer that is probably not this PR's fault.

Reply:
Maybe you've seen it, but in case not, let me mention that I tried to add shortcut instances over subobjects (IntermediateField in this case) in #9291 and they have been effective in avoiding the timeouts or manual `haveI/letI`s.